Unbreak examples after glutin update

Some examples were broken by #1757, they should be fixed. This is a blocker for the 0.26.0 release.

- [x] Tutorial examples (https://github.com/glium/glium/pull/1765)
- [ ] deferred: needs various ownership changes. (#1771)
- [x] fullscreen: needs access to EventLoopWindowTarget, which is not passed in by the current `start_loop`. (#1770)
- [x] screenshot-asynchronous: uses a `loop` for game loop, instead of `start_loop` (#1769)
- [x] shadow_mapping: uses a `loop` for game loop, instead of `start_loop` (#1769)

@est31 I am trying to run the instancing example in a new project, so maybe that is my issue. I copied the code for `instancing.rs` into `main.rs`, added `glium` to my `Cargo.toml` file, copy pasted the support module into my `src` folder, and get a bunch of errors:

Cargo.toml:
```
[dependencies]
glium = "*"
genmesh = "0.6"
obj = { version = "0.9", features = ["genmesh"] }
rand = "0.7"
glutin = "0.22.0-alpha1"
```

src/main.rs: (same as glium/examples/instances.rs)
src/support: (same as glium/examples/support)


Errors:

```
$ cargo run
   Compiling instances v0.1.0 (/home/sam/workspace/instances)
error[E0432]: unresolved import `glium::glutin::event_loop`
 --> src/support/mod.rs:9:20
  |
9 | use glium::glutin::event_loop::{EventLoop, ControlFlow};
  |                    ^^^^^^^^^^ could not find `event_loop` in `glutin`

error[E0432]: unresolved import `glium::glutin::event`
  --> src/support/mod.rs:10:20
   |
10 | use glium::glutin::event::{Event, StartCause};
   |                    ^^^^^ could not find `event` in `glutin`

error[E0433]: failed to resolve: could not find `event_loop` in `glutin`
  --> src/main.rs:17:30
   |
17 |     let event_loop = glutin::event_loop::EventLoop::new();
   |                              ^^^^^^^^^^ could not find `event_loop` in `glutin`

error[E0433]: failed to resolve: could not find `window` in `glutin`
  --> src/main.rs:18:22
   |
18 |     let wb = glutin::window::WindowBuilder::new();
   |                      ^^^^^^ could not find `window` in `glutin`

error[E0433]: failed to resolve: could not find `event` in `glutin`
   --> src/main.rs:124:25
    |
124 |                 glutin::event::Event::WindowEvent { event, .. } => match event {
    |                         ^^^^^ could not find `event` in `glutin`

error[E0433]: failed to resolve: could not find `event` in `glutin`
   --> src/main.rs:125:29
    |
125 |                     glutin::event::WindowEvent::CloseRequested => action = support::Action::Stop,
    |                             ^^^^^ could not find `event` in `glutin`

error: aborting due to 6 previous errors

Some errors have detailed explanations: E0432, E0433.
For more information about an error, try `rustc --explain E0432`.
error: Could not compile `instances`.

To learn more, run the command again with --verbose.
```

@est31 I still get the following:

```
$ cargo run
   Compiling instances v0.1.0 (/home/sam/workspace/instances)
error[E0433]: failed to resolve: could not find `event` in `glutin`
   --> src/support/camera.rs:153:53
    |
153 |     pub fn process_input(&mut self, event: &glutin::event::WindowEvent) {
    |                                                     ^^^^^ could not find `event` in `glutin`

error[E0433]: failed to resolve: could not find `event` in `glutin`
   --> src/support/camera.rs:155:21
    |
155 |             glutin::event::WindowEvent::KeyboardInput { input, .. } => input,
    |                     ^^^^^ could not find `event` in `glutin`

error[E0433]: failed to resolve: could not find `event` in `glutin`
   --> src/support/camera.rs:158:46
    |
158 |         let pressed = input.state == glutin::event::ElementState::Pressed;
    |                                              ^^^^^ could not find `event` in `glutin`

error[E0433]: failed to resolve: could not find `event` in `glutin`
   --> src/support/camera.rs:164:21
    |
164 |             glutin::event::VirtualKeyCode::Up => self.moving_up = pressed,
    |                     ^^^^^ could not find `event` in `glutin`

error[E0433]: failed to resolve: could not find `event` in `glutin`
   --> src/support/camera.rs:165:21
    |
165 |             glutin::event::VirtualKeyCode::Down => self.moving_down = pressed,
    |                     ^^^^^ could not find `event` in `glutin`

error[E0433]: failed to resolve: could not find `event` in `glutin`
   --> src/support/camera.rs:166:21
    |
166 |             glutin::event::VirtualKeyCode::A => self.moving_left = pressed,
    |                     ^^^^^ could not find `event` in `glutin`

error[E0433]: failed to resolve: could not find `event` in `glutin`
   --> src/support/camera.rs:167:21
    |
167 |             glutin::event::VirtualKeyCode::D => self.moving_right = pressed,
    |                     ^^^^^ could not find `event` in `glutin`

error[E0433]: failed to resolve: could not find `event` in `glutin`
   --> src/support/camera.rs:168:21
    |
168 |             glutin::event::VirtualKeyCode::W => self.moving_forward = pressed,
    |                     ^^^^^ could not find `event` in `glutin`

error[E0433]: failed to resolve: could not find `event` in `glutin`
   --> src/support/camera.rs:169:21
    |
169 |             glutin::event::VirtualKeyCode::S => self.moving_backward = pressed,
    |                     ^^^^^ could not find `event` in `glutin`

error: aborting due to 9 previous errors

For more information about this error, try `rustc --explain E0433`.
error: Could not compile `instances`.

To learn more, run the command again with --verbose.
```

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -49,6 +49,7 @@ fnv = "1.0.5"
 gl_generator = "0.13"
 
 [dev-dependencies]
+rental = "0.5.4"
 cgmath = "0.17"
 genmesh = "0.6"
 image = "0.21"

--- a/examples/deferred.rs
+++ b/examples/deferred.rs
@@ -2,6 +2,8 @@
 extern crate glium;
 extern crate cgmath;
 extern crate image;
+#[macro_use]
+extern crate rental;
 
 use glium::index::PrimitiveType;
 #[allow(unused_imports)]
@@ -9,6 +11,23 @@ use glium::{glutin, Surface};
 use std::io::Cursor;
 
 mod support;
+
+pub struct Dt {
+    depthtexture: glium::texture::DepthTexture2d,
+    textures: [glium::texture::Texture2d; 4],
+    light_texture: glium::texture::Texture2d,
+}
+
+rental! {
+    mod my_rentals {
+        use super::Dt;
+        #[rental]
+        pub struct Data {
+            dt: Box<Dt>,
+            buffs: (glium::framebuffer::MultiOutputFrameBuffer<'dt>, glium::framebuffer::SimpleFrameBuffer<'dt>, &'dt Dt),
+        }
+    }
+}
 
 fn main() {
     use cgmath::SquareMatrix;
@@ -264,11 +283,22 @@ fn main() {
     let texture3 = glium::texture::Texture2d::empty_with_format(&display, glium::texture::UncompressedFloatFormat::F32F32F32F32, glium::texture::MipmapsOption::NoMipmap, 800, 500).unwrap();
     let texture4 = glium::texture::Texture2d::empty_with_format(&display, glium::texture::UncompressedFloatFormat::F32F32F32F32, glium::texture::MipmapsOption::NoMipmap, 800, 500).unwrap();
     let depthtexture = glium::texture::DepthTexture2d::empty_with_format(&display, glium::texture::DepthFormat::F32, glium::texture::MipmapsOption::NoMipmap, 800, 500).unwrap();
-    let output = &[("output1", &texture1), ("output2", &texture2), ("output3", &texture3), ("output4", &texture4)];
-    let mut framebuffer = glium::framebuffer::MultiOutputFrameBuffer::with_depth_buffer(&display, output.iter().cloned(), &depthtexture).unwrap();
-
     let light_texture = glium::texture::Texture2d::empty_with_format(&display, glium::texture::UncompressedFloatFormat::F32F32F32F32, glium::texture::MipmapsOption::NoMipmap, 800, 500).unwrap();
-    let mut light_buffer = glium::framebuffer::SimpleFrameBuffer::with_depth_buffer(&display, &light_texture, &depthtexture).unwrap();
+
+    let mut tenants = my_rentals::Data::new(
+        Box::new(Dt {
+            depthtexture,
+            textures: [texture1, texture2, texture3, texture4],
+            light_texture,
+        }),
+        |dt| {
+            let output = [("output1", &dt.textures[0]), ("output2", &dt.textures[1]), ("output3", &dt.textures[2]), ("output4", &dt.textures[3])];
+            let framebuffer = glium::framebuffer::MultiOutputFrameBuffer::with_depth_buffer(&display, output.into_iter().cloned(), &dt.depthtexture).unwrap();
+            let light_buffer = glium::framebuffer::SimpleFrameBuffer::with_depth_buffer(&display, &dt.light_texture, &dt.depthtexture).unwrap();
+            (framebuffer, light_buffer, dt)
+        }
+    );
+
 
     let ortho_matrix: cgmath::Matrix4<f32> = cgmath::ortho(0.0, 800.0, 0.0, 500.0, -1.0, 1.0);
 
@@ -308,70 +338,72 @@ fn main() {
 
     // the main loop
     support::start_loop(event_loop, move |events| {
-        // prepass
-        let uniforms = uniform! {
-            perspective_matrix: Into::<[[f32; 4]; 4]>::into(perspective_matrix),
-            view_matrix: Into::<[[f32; 4]; 4]>::into(view_matrix),
-            model_matrix: Into::<[[f32; 4]; 4]>::into(model_matrix),
-            tex: &opengl_texture
-        };
-        framebuffer.clear_color(0.0, 0.0, 0.0, 0.0);
-        framebuffer.draw(&floor_vertex_buffer, &floor_index_buffer, &prepass_program, &uniforms, &Default::default()).unwrap();
+        tenants.rent_mut(|(framebuffer, light_buffer, dt)| {
+            // prepass
+            let uniforms = uniform! {
+                perspective_matrix: Into::<[[f32; 4]; 4]>::into(perspective_matrix),
+                view_matrix: Into::<[[f32; 4]; 4]>::into(view_matrix),
+                model_matrix: Into::<[[f32; 4]; 4]>::into(model_matrix),
+                tex: &opengl_texture
+            };
+            framebuffer.clear_color(0.0, 0.0, 0.0, 0.0);
+            framebuffer.draw(&floor_vertex_buffer, &floor_index_buffer, &prepass_program, &uniforms, &Default::default()).unwrap();
 
-        // lighting
-        let draw_params = glium::DrawParameters {
-            //depth_function: glium::DepthFunction::IfLessOrEqual,
-            blend: glium::Blend {
-                color: glium::BlendingFunction::Addition {
-                    source: glium::LinearBlendingFactor::One,
-                    destination: glium::LinearBlendingFactor::One
+            // lighting
+            let draw_params = glium::DrawParameters {
+                //depth_function: glium::DepthFunction::IfLessOrEqual,
+                blend: glium::Blend {
+                    color: glium::BlendingFunction::Addition {
+                        source: glium::LinearBlendingFactor::One,
+                        destination: glium::LinearBlendingFactor::One
+                    },
+                    alpha: glium::BlendingFunction::Addition {
+                        source: glium::LinearBlendingFactor::One,
+                        destination: glium::LinearBlendingFactor::One
+                    },
+                    constant_value: (1.0, 1.0, 1.0, 1.0)
                 },
-                alpha: glium::BlendingFunction::Addition {
-                    source: glium::LinearBlendingFactor::One,
-                    destination: glium::LinearBlendingFactor::One
-                },
-                constant_value: (1.0, 1.0, 1.0, 1.0)
-            },
-            .. Default::default()
-        };
-        light_buffer.clear_color(0.0, 0.0, 0.0, 0.0);
-        for light in lights.iter() {
+                .. Default::default()
+            };
+            light_buffer.clear_color(0.0, 0.0, 0.0, 0.0);
+            for light in lights.iter() {
+                let uniforms = uniform! {
+                    matrix: Into::<[[f32; 4]; 4]>::into(ortho_matrix),
+                    position_texture: &dt.textures[0],
+                    normal_texture: &dt.textures[1],
+                    light_position: light.position,
+                    light_attenuation: light.attenuation,
+                    light_color: light.color,
+                    light_radius: light.radius
+                };
+                light_buffer.draw(&quad_vertex_buffer, &quad_index_buffer, &lighting_program, &uniforms, &draw_params).unwrap();
+            }
+
+            // composition
             let uniforms = uniform! {
                 matrix: Into::<[[f32; 4]; 4]>::into(ortho_matrix),
-                position_texture: &texture1,
-                normal_texture: &texture2,
-                light_position: light.position,
-                light_attenuation: light.attenuation,
-                light_color: light.color,
-                light_radius: light.radius
+                decal_texture: &dt.textures[2],
+                lighting_texture: &dt.light_texture
             };
-            light_buffer.draw(&quad_vertex_buffer, &quad_index_buffer, &lighting_program, &uniforms, &draw_params).unwrap();
-        }
+            let mut target = display.draw();
+            target.clear_color(0.0, 0.0, 0.0, 0.0);
+            target.draw(&quad_vertex_buffer, &quad_index_buffer, &composition_program, &uniforms, &Default::default()).unwrap();
+            target.finish().unwrap();
 
-        // composition
-        let uniforms = uniform! {
-            matrix: Into::<[[f32; 4]; 4]>::into(ortho_matrix),
-            decal_texture: &texture3,
-            lighting_texture: &light_texture
-        };
-        let mut target = display.draw();
-        target.clear_color(0.0, 0.0, 0.0, 0.0);
-        target.draw(&quad_vertex_buffer, &quad_index_buffer, &composition_program, &uniforms, &Default::default()).unwrap();
-        target.finish().unwrap();
+            let mut action = support::Action::Continue;
 
-        let mut action = support::Action::Continue;
-
-        // polling and handling the events received by the window
-        for event in events {
-            match event {
-                glutin::event::Event::WindowEvent { event, .. } => match event {
-                    glutin::event::WindowEvent::CloseRequested => action = support::Action::Stop,
+            // polling and handling the events received by the window
+            for event in events {
+                match event {
+                    glutin::event::Event::WindowEvent { event, .. } => match event {
+                        glutin::event::WindowEvent::CloseRequested => action = support::Action::Stop,
+                        _ => (),
+                    },
                     _ => (),
-                },
-                _ => (),
-            }
-        };
+                }
+            };
 
-        action
+            action
+        })
     });
 }

